File: android/src/main/kotlin/noise/protocol/KeyPair.kt
package noise.protocol

/**
 * Represents a Diffie-Hellman key pair containing raw private and public key bytes.
 *
 * Key material is stored as raw byte arrays without ASN.1 or other encoding wrappers.
 * The key sizes depend on the DH function in use (32 bytes for Curve25519, 56 bytes for X448).
 *
 * **Important:** Because this class holds sensitive private key material, callers should
 * zero out [privateKey] when it is no longer needed to reduce the window of exposure.
 *
 * Uses content-based equality rather than reference equality for [ByteArray] fields.
 *
 * @property privateKey The raw private key bytes.
 * @property publicKey The raw public key bytes.
 * @see DH
 * @see SecureBuffer
 */
data class KeyPair(val privateKey: ByteArray, val publicKey: ByteArray) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is KeyPair) return false
        return privateKey.contentEquals(other.privateKey) && publicKey.contentEquals(other.publicKey)
    }

    override fun hashCode(): Int = 31 * privateKey.contentHashCode() + publicKey.contentHashCode()
}

File: ios/Sources/NoiseProtocol/KeyPair.swift
import Foundation

/// A cryptographic key pair consisting of a private key and its corresponding public key.
///
/// Used throughout the Noise Protocol for Diffie-Hellman operations. The key sizes depend
/// on the DH function in use (e.g., 32 bytes for Curve25519, 56 bytes for X448).
public struct KeyPair: Sendable {
    /// The private (secret) key material. Must be kept confidential.
    public let privateKey: Data
    /// The public key derived from the private key. Safe to share with peers.
    public let publicKey: Data

    /// Creates a new key pair from raw key data.
    ///
    /// - Parameters:
    ///   - privateKey: The raw private key bytes.
    ///   - publicKey: The raw public key bytes corresponding to the private key.
    public init(privateKey: Data, publicKey: Data) {
        self.privateKey = privateKey
        self.publicKey = publicKey
    }
}
